rfcs/react-components/convergence/theme-shared-colors.md

```diff
@@ -0,0 +1,114 @@
+
+### Real Application Performance
+
+To avoid any performance regressions there are performance gates running as part of partner application CI.
+
+The test case which is interesting for us is WLT hot - that opens a new application window (pre-heat), closes the window and opens it again measuring the performance. The current time for this test is around 700ms.
+
+Just adding v9 `Provider` with a theme, the test time rises by **20ms** (with the original approach having 1200 tokens and alias references it was 80ms) which does not pass the gate. For that reason we have been stuck on this for 3 months already not being able to enable the v9 `Provider` in the application, blocking feature teams in starting migration to v9.
+
+As an experiment we tried to reduce the tokens to 200, the test time degradation was 7ms.
+
+| Theme structure               | WLT degradation [ms] |
+| ----------------------------- | -------------------: |
+| 1200 tokens, alias references |                   70 |
+| 600 tokens                    |                   20 |
+| 200 tokens                    |                    7 |
+
+**The numbers confirm the performance decreases linearly depending on the number of CSS variables (and therefore theme tokens).**
+
+### New Shared Color Alias tokens request
+
+_The numbers might be inaccurate as the feature is not fully spec-ed yet!_
+
+In order to support new scenarios (colorful calendar view), there is a new request to add 3 new shared color alias tokens. Adding them would result in adding 3 alias tokens for each of 49 shared colors = adding 147 new shared color alias tokens which would affect application performance.
+
+## Detailed Design or Proposal
+
+The theme currently contains 49 global shared colors.
+
+Current alias structure forces each ramp to be uniform (the same set of alias tokens for each shared color), resulting in 49 alias tokens each time we need a new line item (new alias token). Today we have 9 alias slots = total alias count is 441 (49x9).
+
+Components have varying degrees of color needs.
+
+Instead of 1 uniform Shared alias table (today's experience), separate out into 3 tables:
+
+1. Semantic States
+   - There are 5 unique global colors needed to support this set (Danger, Success, Severe, Warning, Out of office)
+   - Assume 9 alias tokens (Could potentially remove 2 from this table as they were avatar specific)
+   - 45 alias tokens in total (5 x 9)
+2. Persona
+   - There are 30 unique global colors needed to support this set
+   - Only needs 3 alias tokens
+   - 90 tokens in total (30 x 3)
+3. Category (calendar categories) (not fully defined yet)
+   - 10 unique global colors to support this set
+   - Assuming 9 alias slots (subject to increase)
+   - 90 tokens in total (10 x 9)
+
+Dividing up by 3 tables would reduce token count from 441 to ~225 (Table 1+Table 2+Table 3) while adding support for Calendar scenario.
+
```

Comment:
Two questions:
1. Is it reasonable to break up some of these tokens such that they can be used in a pay for play way? Especially Persona and Category. For most of our apps they will probably need them, but I could definitely imagine scenarios where applications do not need those colors and therefore could save token count.
	
2. As a follow on, even if it doesn't make sense for these tokens to be optional. Should we consider adding new sets of tokens in an additive way? Currently if we have 1 monolithic set of tokens, I suspect we'll be in trouble in a few years after we continue to add more tokens for various scenarios.
Allowing optional tokens does increase complexity, so perhaps there is a way to "load all" tokens if you want simplicity, but giving applications knobs to allow them to avoid loading all tokens on initial app launch may be useful. We should vette this ideas with products, but something to consider.

Reply:
As mentioned bellow: _There is still a longer term plan to explore partial themes, theme splitting and lazy loading._

**partial theme** is exactly what you are proposing:

#### Use partial theme in the application
The current theme (`master`, Apr 27, 2022) contains 623 tokens. Out of that 600 tokens, only ~200 tokens are used by the library.

Instead of using the full theme, application can only inject the tokens in use.

This can be done manually, or automatically during build time.

##### 👍 Pros
- significantly reduces the number of tokens -> improves WLT

###### 👎 Cons
- manual way is hard to maintain and error prone
- automated way is complex (how about processed styles in dependencies - "manifests"?)
- hard for parts (shared components) which are not part of the application bundle (OTA)

#### Theme splitting, lazy loading
Similar to previous approach, but find a way to "lazy inject" (perhaps not lazy load, all the tokens are part of the initial bundle?) more tokens when needed.

##### 👍 Pros
- significantly reduces the number of tokens -> improves WLT
- any token can be used

###### 👎 Cons
- complex, a lot of unknowns
- adding variables would cause reflows -> perf hit later

Reply:
Thanks for the breakdown, this makes sense. 

"For partial theme in the application". A way to mitigate "manual way is hard" is kind of what we're doing with the 3 separate tables or having token sets grouped by logical use. For instance, "Categories" or "Persona" have a particular scenario they are used for. So, you would include *all* Categories or none at all, rather than have to decide which categories you use. 

Of course, this doesn't address dynamically loaded experiences and is still tricky for shared components, but it would be easier to know/document that if you are using Avatar it requires the Persona token set.

Reply:
Added to the RFC.